core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java

```diff
@@ -0,0 +1,245 @@
+
+    public GlobalCheckpointService(ShardId shardId, IndexSettings indexSettings) {
+        super(shardId, indexSettings);
+        activeLocalCheckpoints = new ObjectLongHashMap<>(1 + indexSettings.getNumberOfReplicas());
+        inSyncLocalCheckpoints = new ObjectLongHashMap<>(indexSettings.getNumberOfReplicas());
+        trackingLocalCheckpoint = new ObjectLongHashMap<>(indexSettings.getNumberOfReplicas());
+    }
+
+
+    /**
+     * notifies the service of a local checkpoint. if the checkpoint is lower than the currently known one,
+     * this is a noop. Last, if the allocation id is not yet known, it is ignored. This to prevent late
+     * arrivals from shards that are removed to be re-added.
+     */
+    synchronized public void updateLocalCheckpoint(String allocationId, long localCheckpoint) {
+        if (updateLocalCheckpointInMap(allocationId, localCheckpoint, activeLocalCheckpoints, "active")) {
+            return;
+        }
+        if (updateLocalCheckpointInMap(allocationId, localCheckpoint, inSyncLocalCheckpoints, "inSync")) {
+            return;
+        }
+        if (updateLocalCheckpointInMap(allocationId, localCheckpoint, trackingLocalCheckpoint, "tracking")) {
+            return;
+        }
+        logger.trace("local checkpoint of [{}] ([{}]) wasn't found in any map. ignoring.", allocationId, localCheckpoint);
+    }
+
+    private boolean updateLocalCheckpointInMap(String allocationId, long localCheckpoint,
+                                               ObjectLongMap<String> checkpointsMap, String name) {
+        assert Thread.holdsLock(this);
+        int indexOfKey = checkpointsMap.indexOf(allocationId);
+        if (indexOfKey < 0) {
+            return false;
+        }
+        long current = checkpointsMap.indexGet(indexOfKey);
+        // nocommit: this can change when we introduces rollback/resync
+        if (current < localCheckpoint) {
+            checkpointsMap.indexReplace(indexOfKey, localCheckpoint);
+            if (logger.isTraceEnabled()) {
+                logger.trace("updated local checkpoint of [{}] to [{}] (type [{}])", allocationId, localCheckpoint,
+                    name);
+            }
+        } else {
+            logger.trace("skipping update local checkpoint [{}], current check point is higher " +
+                "(current [{}], incoming [{}], type [{}])",
+                allocationId, current, localCheckpoint, allocationId);
+        }
+        return true;
+    }
+
+    /**
+     * Scans through the currently known local checkpoints and updates the global checkpoint accordingly.
+     *
+     * @return true if the checkpoint has been updated or if it can not be updated since one of the local checkpoints
+     * of one of the active allocations is not known.
+     */
+    synchronized public boolean updateCheckpointOnPrimary() {
+        long minCheckpoint = Long.MAX_VALUE;
+        if (activeLocalCheckpoints.isEmpty() &&  inSyncLocalCheckpoints.isEmpty()) {
```

Comment:
Nit: there is an extra space after the `&&` and before `inSyncLocalCheckpoints`.